utils/babeltrace_thapi.in

```diff
@@ -120,96 +127,78 @@ if !$options[:live]
+  require 'babeltrace_hip_lib' if $options[:backends].include?('hip')
+
+  consume = lambda { |iterator, _|
+    iterator.next_messages.each do |m|
+      next unless m.type == :BT_MESSAGE_TYPE_EVENT
+
+      e = m.event
+      l = $event_lambdas[e.name]
+      next unless l || !$options[:restrict]
+
+      str = "#{Time.at(0, m.get_default_clock_snapshot.ns_from_origin, :nsec).strftime('%H:%M:%S.%9L')}"
+      if $options[:context]
+        str << " - #{e.stream.trace.get_environment_entry_value_by_name('hostname')}"
+        str << ' - ' << e.get_common_context_field.value.collect do |k, v|
+          "#{k}: #{v}"
+        end.join(', ')
       end
-    }
-    graph.add_simple_sink('babeltrace_thapi', consume)
-  when 'tally'
-
-    thapi_plugins.each do |plugin|
-      interval_plugins << plugin.get_filter_component_class_by_name('aggreg') if plugin.name == 'btx_aggreg'
-    end
-
-    comp_intervals = interval_plugins.map.with_index do |comp, i|
-      graph.add_component(comp, "interval_#{i}")
+      str << " - #{e.name}: "
+      str << (l ? l.call(e.payload_field.value) : e.payload_field.to_s)
+      puts str
     end
-
-    # Only one, but for generality return an array
-    sink_plugin = thapi_plugins.filter_map do |plugin|
-      plugin.get_sink_component_class_by_name('tally') if plugin.name == 'btx_tally'
-    end.first
-
-    comp_tally = graph.add_component(sink_plugin, 'tally',
-                                     params: $options_tally.transform_values { |_, v| v } )
-    [comp_intervals, comp_tally]
-  when 'timeline'
-    comp_intervals = interval_plugins.map.with_index do |comp, i|
-      graph.add_component(comp, "interval_#{i}")
-    end
-
-    # Only one, but for generality return an array
-    sink_plugin = thapi_plugins.filter_map do |plugin|
-      plugin.get_sink_component_class_by_name('timeline') if plugin.name == 'btx_timeline'
-    end.first
-
-    comp_timeline = graph.add_component(sink_plugin, 'timeline')
-    [comp_intervals, comp_timeline]
+  }
+  comps << graph.add_simple_sink('babeltrace_thapi', consume)
+when 'tally'
+  $options[:backends].each_with_index do |name, i|
+    comps << graph.add_component(components_classes["filter.#{name}interval.interval"], "interval_#{i}")
   end
-
-# Sources to muxer
-comp_sources.flat_map(&:output_ports).each_with_index do |op, i|
-  ip = comp_muxer.input_port(i)
-  graph.connect_ports(op, ip)
+  comps << graph.add_component(components_classes['filter.btx_aggreg.aggreg'], 'aggreg')
+  comps << graph.add_component(components_classes['sink.btx_tally.tally'], 'tally',
+                              params: $options_tally.transform_values { |_, v| v })
+when 'timeline'
+  $options[:backends].each_with_index do |name, i|
+    comps << graph.add_component(components_classes["filter.#{name}interval.interval"], "interval_#{i}")
+  end
+  comps << graph.add_component(components_classes['sink.btx_timeline.timeline'], 'timeline')
 end
```

Comment:
Do we want an else here, and raise an error if we reach it? Or are the arguments sanitized above?

Reply:
We have an error but it's ugly. It's `Cannot initialize a null component` . ^^
Yes, I can create a custom `.fetch` method that raises a nice error message if not.
 In theory, it can never happen but doesn't cost that much to be clean.

Reply:
Sounds good.

Reply:
Use fetch, as `keyerror` is good enough as error message.  ( 03dd4e76ab55724f9ff8e72d50c0fc30526466c2 )